src/runtime/dsl.ts

```diff
@@ -0,0 +1,431 @@
+      }
+      this.variableLookup[b.id] = [];
+    } else if(aValue !== bValue) {
+      throw new Error(`Trying to equivalence two static values that aren't the same: ${aValue} and ${bValue}`);
+    }
+  }
+
+  prepare() {
+    let functions = this.functions.slice() as (DSLFunction | undefined)[];
+    // look through the functions for equivalences before we do anything else.
+    let ix = 0;
+    for(let func of functions) {
+      if(!func || func.path[1] !== "==") continue;
+      this.equivalence(func.args[0], func.args[1]);
+      functions[ix] = undefined;
+      ix++;
+    }
+    let registerIx = 0;
+    for(let id in this.variableLookup) {
+      let variable = this.variableLookup[id][0] as DSLVariable;
+      if(!variable || !isRegister(variable.value)) continue;
+      if(variable.value.offset !== -1) throw new Error("We've somehow already assigned a variable's register");
+      variable.value.offset = registerIx++;
+    }
+    let items = functions.concat(this.records as any[]);
+    let constraints = [];
+    let nodes = [];
+    for(let toCompile of items) {
+      if(!toCompile) continue;
+      let compiled = toCompile.compile();
+      if(!compiled) continue;
+      for(let item of compiled) {
+        if(item instanceof Scan || item instanceof FunctionConstraint) {
+          constraints.push(item);
+        } else {
+          nodes.push(item as Node);
+        }
+      }
+    }
+    nodes.unshift(new runtime.JoinNode(constraints))
+    this.block = new runtime.Block(this.name, nodes);
+    console.log(this);
+  }
+
+  //-------------------------------------------------------------------
+  // Code transforms
+  //--------------------------------------------------------------------
+
+  transformBlockCode(code:string, functionArgs:string[]):string {
+
+    let libArg = functionArgs[2];
+    let hasChanged = true;
+    let infixParam = "((?:(?:[a-z0-9_\.]+(?:\\[\".*?\"\\])?)+(?:\\(.*\\))?)|\\(.*\\))";
+    let stringPlaceholder = "(____[0-9]+____)";
+
+    let strings:string[] = [];
+    code = code.replace(/"(?:[^"\\]|\\.)*"/gi, function(str) {
+      strings.push(str);
+      return "____" + (strings.length - 1) + "____";
+    })
+
+    let stringAddition = new RegExp(`(?:${infixParam}\\s*\\+\\s*${stringPlaceholder})|(?:${stringPlaceholder}\\s*\\+\\s*${infixParam})`,"gi");
+    hasChanged = true;
+    while(hasChanged) {
+      hasChanged = false;
+      code = code.replace(stringAddition, (str, left, right, left2, right2) => {
+        hasChanged = true;
+        if(left === undefined) {
+          left = left2;
+          right = right2;
+        }
+        left = this.transformBlockCode(left, functionArgs);
+        right = this.transformBlockCode(right, functionArgs);
+        strings.push(`${libArg}.eve.internal.concat(${left}, ${right})`);
+        return "____" + (strings.length - 1) + "____";
+      })
+    }
+
+    let multiply = new RegExp(`${infixParam}\\s*(\\*|\\/)\\s*${infixParam}`, "gi");
```

Comment:
Since this isn't in a hot path, can we maybe make a helper function to cut the repetition down a bit? The extra noise makes it a bit hard for me to wade through.